servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyPipelinedConnection.java

```diff
@@ -0,0 +1,344 @@
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.concurrent.internal.ConcurrentUtils;
+import io.servicetalk.transport.api.ExecutionContext;
+import io.servicetalk.transport.netty.internal.FlushStrategy;
+import io.servicetalk.transport.netty.internal.NettyConnection;
+import io.servicetalk.transport.netty.internal.NettyConnectionContext;
+import io.servicetalk.transport.netty.internal.WriteDemandEstimator;
+import io.servicetalk.transport.netty.internal.WriteDemandEstimators;
+
+import io.netty.channel.Channel;
+
+import java.net.SocketAddress;
+import java.net.SocketOption;
+import java.util.Queue;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+import javax.net.ssl.SSLSession;
+
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static io.servicetalk.concurrent.internal.ConcurrentUtils.acquirePendingLock;
+import static io.servicetalk.concurrent.internal.ConcurrentUtils.releasePendingLock;
+import static io.servicetalk.concurrent.internal.SubscriberUtils.deliverTerminalFromSource;
+import static io.servicetalk.utils.internal.PlatformDependent.newUnboundedMpscQueue;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.atomic.AtomicIntegerFieldUpdater.newUpdater;
+
+/**
+ * Contract for using a {@link NettyConnection} to make pipelined requests, typically for a client.
+ * <p>
+ * Pipelining allows to have concurrent requests processed on the server but still deliver responses in order.
+ * This eliminates the need for request-response correlation, at the cost of head-of-line blocking.
+ * @param <Req> Type of requests sent on this connection.
+ * @param <Resp> Type of responses read from this connection.
+ */
+final class NettyPipelinedConnection<Req, Resp> implements NettyConnectionContext {
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<NettyPipelinedConnection> writeQueueLockUpdater =
+            newUpdater(NettyPipelinedConnection.class, "writeQueueLock");
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<NettyPipelinedConnection> readQueueLockUpdater =
+            newUpdater(NettyPipelinedConnection.class, "readQueueLock");
+    private final NettyConnection<Resp, Req> connection;
+    private final Queue<WriteTask> writeQueue;
+    private final Queue<Subscriber<? super Resp>> readQueue;
+    @SuppressWarnings("unused")
+    private volatile int writeQueueLock;
+    @SuppressWarnings("unused")
+    private volatile int readQueueLock;
+
+    /**
+     * New instance.
+     *
+     * @param connection {@link NettyConnection} requests to which are to be pipelined.
+     */
+    NettyPipelinedConnection(NettyConnection<Resp, Req> connection) {
+        this.connection = requireNonNull(connection);
+        writeQueue = newUnboundedMpscQueue();
+        readQueue = newUnboundedMpscQueue();
```

Comment:
nit: I wonder if unbounded may be risky here... But I guess it should be fine ?

Reply:
We apply limits at a higher level in the client.